### ISSUE TITLE
Can a user later decline (and thus request deletion) of their personal data based on the previously accepted cookie?

### ISSUE THREAD

**Issue by sealad886** [django-commons/django-cookie-consent#131]
Hi there, just trying to wrap my head around the intricacies of GDPR here. My understanding is that we would have to remove a user's personal data on the back end (within reason) if they later withdrew consent. Is there a way to do that? I didn't see that immediately obvious. Thanks so much!

**Comment by some1ataplace** [comment]
There are a couple of ways to handle this, but the situation is complicated. 

I am not a lawyer, but for legal purposes, I think there needs to be some sort of logging of the cookie acceptances/declines most likely by user IP address (to handle both authenticated and anonymous users). Seems like all the major players in the Cookie Consent space track every action with a new record each time. 

Keep in mind, IP addresses are considered to be Personally Identifiable Info (PII) and that should probably be de-identified using some sort of scrambling or encryption/decryption techniques.

You can add a "Do Not Sell My Personal Information" form to your site. The user could then submit that form with the following options: Request Data, Delete Data, Update Data, Do Not Sell Data. You would have to track all authenticated and anonymous user "Do Not Sell My Personal Information" submissions using the user's IP address. If the user who is submitting a Do Not Sell My Personal Information form's IP address matches a Cookie Log record and their form specified they want to Delete Data, then you can probably delete the Cookie Log record. But now, you will have a Do Not Sell My Personal Information record (unless you delete that record too). Maybe you can ask the user what exactly they want to Delete and automatically delete that thing they specified, then delete the Do Not Sell My Personal information record afterwards. 

If the user has an account with your site, give them the option to delete their entire account, and CASCADE all their database data when their account deletion completes. If the user who is deleting their account's IP address matches any of the Cookie Log or Do Not Sell My Personal Information records, then you can probably delete those too. 

An issue arises if you have multiple cookie groups. What happens if a user only declines one of those groups and not all the other groups? Maybe you can delete that one Cookie Log group record, but the other group records will still remain. If a user declines every cookie group, then you can probably delete all the Cookie Log records for that IP address. 

Perhaps the best way to handle this is to Update a single IP address Cookie Log database record each time a user accepts/declines. That way, you have the best of both worlds - cookie logging and not flooding your database with tons of records of every single action the user makes in regards to accepting/declining cookies. 

You can of course turn cookie logging off completely, but do so at your own risk. You can also setup a django management command or a cron job to delete old Cookie Log records after a certain period of time, but that probably is not a good idea either from a legal standpoint. 

If anyone else has any other better ideas on how to handle this, I am sure we would all be interested in hearing from you. 


```
class CookieLog(models.Model):
    ip_address = models.GenericIPAddressField()
    action = models.CharField('Action', max_length=256) #most recent action of accept/decline
    cookie_value = models.CharField('Cookie Value', max_length=256) #most recent cookie value
    date = models.DateTimeField('Date', auto_now_add=True, blank=True)
    groups_accepted = models.CharField('groups_accepted', max_length=256, blank=True) #Track each group that was ever accepted. If a new group gets accepted, add it to this list
    groups_declined = models.CharField('groups_declined', max_length=256, blank=True) #Track each group that was ever declined. If a new group gets declined, add it to this list
    source = models.CharField('source', max_length=256, blank=True) #Cookie Bar, Cookie Page Button



#Declines
if settings.COOKIE_CONSENT_LOGGING == 1:
    ip = str(get_client_ip(request))
    if CookieLog.objects.filter(ip_address=ip).exists():
        CookieLog.objects.filter(ip_address=ip).update(action='Declined',groups_declined=groups_declined,source=source,cookie_value=cookie_declined_string, date=timezone.now())
    else:
        CookieLog.objects.create(ip_address=ip,action='Declined',groups_declined=groups_declined,source=source,cookie_value=cookie_declined_string, date=timezone.now())

#Accepts
 if settings.COOKIE_CONSENT_LOGGING == 1:
    ip = str(get_client_ip(request))
    if CookieLog.objects.filter(ip_address=ip).exists():
        CookieLog.objects.filter(ip_address=ip).update(action='Accepted',groups_accepted=groups_accepted,source=source,cookie_value=cookie_accepted_string, date=timezone.now())
    else:
        CookieLog.objects.create(ip_address=ip,action='Accepted',groups_accepted=groups_accepted,source=source,cookie_value=cookie_accepted_string, date=timezone.now())



class Command(BaseCommand):
    help = 'Delete records in the cookie log table that have been there after a certain time period.'

    def handle(self, *args, **kwargs):
        #Cookie log records in the database between 1-2 days
        recent_cookie_logs = CookieLog.objects.filter(date__gt=timezone.now() - timezone.timedelta(days=2), date__lt=timezone.now() - timezone.timedelta(days=1))
        for x in recent_cookie_logs:
            x.delete()
        #Cookie log records in the database for more than 1 day
        older_cookie_logs = CookieLog.objects.filter(date__lt=timezone.now() - timezone.timedelta(days=1))
        for y in older_cookie_logs:
            y.delete()

```

**Comment by sergei-maertens** [comment]
> Hi there, just trying to wrap my head around the intricacies of GDPR here. My understanding is that we would have to remove a user's personal data on the back end (within reason) if they later withdrew consent. Is there a way to do that? I didn't see that immediately obvious. Thanks so much!

I wouldn't read the GDPR that way - if I understand correctly (but I'm not a lawyer), your project/site needs to have some sort of process for when someone asks for their data to be removed. That can be entirely manual. Rejecting cookie consent does not imply a request for data removal - rather it means that you can no longer track them for analytics through cookies, so those scripts/tools need to deactivate.

So long story short - the scope of django-cookie-consent is not to facilitate user data removal.

**Comment by sergei-maertens** [comment]
I'm going to close this with my above reaction in mind. If this is a terrible mistake on my part, feel free to re-open.
